text/0000-command-overrides.md

```diff
@@ -0,0 +1,132 @@
+# Meta
+[meta]: #meta
+- Name: Default command arguments and overrides
+- Start Date: (fill in today's date: 2021-03-23)
+- Author(s): [@samj1912](https://github.com/samj1912)
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: (put "N/A" unless this replaces an existing RFC, then link to that RFC)
+
+# Summary
+[summary]: #summary
+
+Buildpacks API currently doesn't provide a way to set a default set of arguments for the various process types that can be over-ridden by users. The current behaviour of the laucher is to [append the arguments](https://github.com/buildpacks/spec/blob/main/buildpack.md#launch) provided by the user to the existing list of arguments defined for a process type. The [OCI image spec](https://github.com/opencontainers/image-spec/blob/master/config.md#properties), [Docker](https://docs.docker.com/engine/reference/builder/#understand-how-cmd-and-entrypoint-interact) and [Kubernetes Container specification](https://kubernetes.io/docs/tasks/inject-data-application/define-command-argument-container/#notes) all have support for a construct to allow users to override default arguments. This proposal suggests a way we could achieve something for processes defined by the Buildpacks API. 
+
+# Motivation
+[motivation]: #motivation
+
+- Why should we do this?
+
+This addition would bring Buildpack API in parity with the OCI image spec and what is provided by Dockerfile. It would also enable use cases for overriding default arguments during runtime in the Kubernetes world. In order to achieve something like this, users have resolved to creating shims which implement this functionality. I believe that we could enable a lot more use cases out of the box without users having to implement it themselves.
+
+- What use cases does it support?
+
+  - Providing a default set of arguments
+  - Over-riding these defaults
+
+- What is the expected outcome?
+
+The launcher should be able to determine if arguments have been passed by the user or not and if they have, it should override a default list of arguments by the once specified by the user.
+
+
+# What it is
+[what-it-is]: #what-it-is
+
+The proposal essentially involves adding an additional `additional_default_args` key in the `Launch Metadata` (launch.toml) that is a list of strings. This list of strings would be a set of additional arguments that would be appended to the existing command invocation if the user does not provide any arguments themselves.
+
+```toml
+[[processes]]
+type = "<process type>"
+command = "<command>"
+args = ["<arguments>"]
+additional_default_args = ["<additional>", "<default>", "<arguments>"]
```

Comment:
is there a situation where you'd want to provide _both_ `args` and `additional_default_args`?

Reply:
oh i think i see. `args` _cannot_ be overridden but `additional_default_args` can be?

Reply:
Yes!